plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java

```diff
@@ -0,0 +1,570 @@
+        private final Queue<ExchangeSourceFile> sourceFiles;
+        private final int blockSize;
+        private final int bufferSize;
+
+        @GuardedBy("this")
+        private ExchangeSourceFile currentFile;
+        @GuardedBy("this")
+        private long fileOffset;
+        @GuardedBy("this")
+        private SliceInput sliceInput;
+        @GuardedBy("this")
+        private int sliceSize = -1;
+        private volatile boolean closed;
+        private volatile long bufferRetainedSize;
+        private volatile ListenableFuture<Void> inProgressReadFuture = immediateVoidFuture();
+
+        public AzureExchangeStorageReader(
+                BlobServiceAsyncClient blobServiceAsyncClient,
+                Queue<ExchangeSourceFile> sourceFiles,
+                int blockSize,
+                int maxPageStorageSize)
+        {
+            this.blobServiceAsyncClient = requireNonNull(blobServiceAsyncClient, "blobServiceAsyncClient is null");
+            this.sourceFiles = requireNonNull(sourceFiles, "sourceFiles is null");
+            this.blockSize = blockSize;
+            // Make sure buffer can accommodate at least one complete Slice, and keep reads aligned to block boundaries
+            this.bufferSize = maxPageStorageSize + blockSize;
+
+            // Safe publication of S3ExchangeStorageReader is required as it's a mutable class
+            fillBuffer();
+        }
+
+        @Override
+        public synchronized Slice read()
+                throws IOException
+        {
+            if (closed || !inProgressReadFuture.isDone()) {
+                return null;
+            }
+
+            try {
+                getFutureValue(inProgressReadFuture);
+            }
+            catch (RuntimeException e) {
+                throw new IOException(e);
+            }
+
+            if (sliceSize < 0) {
+                sliceSize = sliceInput.readInt();
+            }
+            Slice data = sliceInput.readSlice(sliceSize);
+
+            if (sliceInput.available() > Integer.BYTES) {
+                sliceSize = sliceInput.readInt();
+                if (sliceInput.available() < sliceSize) {
+                    fillBuffer();
+                }
+            }
+            else {
+                sliceSize = -1;
+                fillBuffer();
+            }
+
+            return data;
+        }
+
+        @Override
+        public ListenableFuture<Void> isBlocked()
+        {
+            // rely on FileSystemExchangeSource implementation to wrap with nonCancellationPropagating
+            return inProgressReadFuture;
+        }
+
+        @Override
+        public synchronized long getRetainedSize()
+        {
+            return INSTANCE_SIZE + bufferRetainedSize;
+        }
+
+        @Override
+        public boolean isFinished()
+        {
+            return closed;
+        }
+
+        @Override
+        public void close()
+        {
+            if (closed) {
```

Comment:
Does the `close` method have to be synchronized?

Reply:
Nice catch, I forgot